### PR TITLE
created different land versions

### PR DESCRIPTION
**Pull request by kmdeck**
1. This recognizes that we will have different types of soil, vegetation, etc models, and so it may be helpful to define an `AbstractSoilModel`, `AbstractVegetationModel`, which inherit from `AbstractModel`. (this may be unnecessary, but it adds some organization, especially once we have `RichardsModel`, `SoilEnergyWaterModel` (or whatever), etc)

2. If we think of different Land Surface Models, with different components in them, as different types of `LandModel`, it might also be useful to introduce an `AbstractLandModel` type and make different concrete types. For example, a `LandHydrologyLSM`, `BiopshereLSM`, `LSM`. Then these types would have preset components, and we wouldnt need the `NotIncluded` types. We hadn't done this before because (IIRC) we wanted to have default behavior for `initialize(land::LandModel)`, `make_update_aux(land::LandModel)`, etc. But we can still do that using `land::AbstractLandModel`.

3. Orthogonal to the above changes, this PR makes it so we `map` over the components of the land model, to automate the `make_ode_function`, `make_update_aux`, and `initialize` methods for land. If we keep all components in a single `LandModel` type, and keep the `NotIncluded` type, this should still work fine.

4. For clarity, I also moved all the abstract model functions to a `models.jl` file, in anticipation of this being a separate module or repo at some point.

TBD on how to finesse `Configurations`

### CODE REVIEW

File: src/ClimaLSM.jl [hunk 1]
```diff
@@ -1,329 +1,121 @@
...
-        Roots.RootsModel{FT}(;
-            configuration = configuration,
-            vegetation_args...,
-        )
-    river = NotIncluded{FT}(:river)
-    snow = NotIncluded{FT}(:snow)
-    carbon = NotIncluded{FT}(:carbon)
-    args = (configuration, soil, vegetation, river, snow, carbon)
-    return LandModel{FT, typeof.(args)...}(args...)
+    #somehow need to pass in soil version and veg version...
```
Comment (jb-mackay) [inline src/ClimaLSM.jl:44]: Right... perhaps adding args for the type of soil & root model to the constructor? Then you would call something like `RootSoilModel{FT}(RichardsModel, soil_args, RootsModel, vegetation_args)`. I think that may be simpler than trying to put the model version into `soil_args` itself.

File: src/ClimaLSM.jl [hunk 2]
```diff
@@ -1,329 +1,121 @@
...
+    #configuration may depend on the versions of the soil and root models too!
+    configuration = Configurations.RootSoilConfiguration{FT}() 
+    soil = Soil.RichardsModel{FT}(; configuration = configuration, soil_args...)
```
Comment (jb-mackay) [inline src/ClimaLSM.jl:47]: Then here we would replace `Soil.RichardsModel` with the method argument that specifies the model version.

File: src/ClimaLSM.jl [hunk 3]
```diff
@@ -1,329 +1,121 @@
...
-    )
+                                     p_interactions...,
+                                     NamedTuple(zip(components, p_state_list))...)
     coords =
-        ClimaCore.Fields.FieldVector(; soil = coords_soil,
-                                     vegetation = coords_vegetation,
-                                     river = coords_river,
-                                     snow = coords_snow,
-                                     carbon = coords_carbon)
+        ClimaCore.Fields.FieldVector(; NamedTuple(zip(components, coords_list))...)
```
Comment (jb-mackay) [inline src/ClimaLSM.jl:83]: Could create a helper function for this since it is used by `Y`, `p`, and `coords` in the same way.

**Review by jb-mackay — APPROVED** [review]
Looks good as a whole, let me know what you think of the other comments but as we discussed, I think this is a natural way to organize things.

File: src/ClimaLSM.jl (hunk 2 above)
Comment (kmdeck) [reply src/ClimaLSM.jl:47]: nice! I replaced this as you suggested like:
```
function RootSoilModel{FT}(;
                           soil_model_type::Type{SM},
                           soil_args::NamedTuple = (;),
                           vegetation_model_type::Type{VM},
                           vegetation_args::NamedTuple = (;),
                           ) where {FT, SM <: AbstractSoilModel{FT}, VM <: AbstractVegetationModel{FT}}
    #configuration may depend on the versions of the soil and root models too
    configuration = Configurations.RootSoilConfiguration{FT}() 
    soil = soil_model_type(; configuration = configuration, soil_args...)
    vegetation = vegetation_model_type(; configuration = configuration,
                                 vegetation_args...,
                                 )
    args = (configuration, soil, vegetation)
    return RootSoilModel{FT, typeof.(args)...}(args...)
end
```

File: src/ClimaLSM.jl [hunk 4]
```diff
@@ -1,329 +1,121 @@
...
+function initialize(land::AbstractLandModel)
+    components = land_components(land)
+    coords_list = map(components) do (component)
+        Domains.coordinates(getproperty(land, component))
+    end
+    Y_state_list = map(zip(components,coords_list)) do (component, coords)
+        getproperty(initialize_prognostic(getproperty(land, component), coords), component)
+    end
+    p_state_list = map(zip(components,coords_list)) do (component, coords)
+        getproperty(initialize_auxiliary(getproperty(land, component), coords), component)
```
[the remark this answers is not in the document]
Comment (kmdeck) [reply src/ClimaLSM.jl:74]: I agree - these map/do blocks are hard to parse. I dont know exactly how mapping over a list of known size compares with a `for` loop over a list of known size. @charleskawczynski  had mentioned to me a few times that appending to a list in a `for` loop is not ideal because it is hard for the compiler (?) to allocate memory efficiently since it doesnt know how long the list will be, but `map` avoids that. Not sure, exactly, since in a `for` loop you also are looping over something of known size. We only need to do this list once, at initialization, so it doesnt matter much, but I was trying to practice that advice.


agreed that all the `get_property` are not nice. Let's revisit this entire `initialize` in a future PR, once we get some of this base functionality merged into main? I agree that for Y we can probably drop the ".component" in standalone mode, but still not sure about `p`.

File: src/ClimaLSM.jl (hunk 3 above)
Comment (kmdeck) [reply src/ClimaLSM.jl:83]: If it's OK with you, I'll wait on doing anything like this as `initialize` might change, and I'm also not sure a helper function would reduce the amount of code overall at this point..

File: src/ClimaLSM.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (charleskawczynski) [reply src/ClimaLSM.jl]: Looping is fine, I think my suggestion was to avoid `push!`ing, which can be expensive (especially for large generic arrays).

[the remark this answers is not in the document]
Comment (kmdeck) [reply src/ClimaLSM.jl]: Hm, I was thinking of this:
```
my_list = []
for x in [1,2,3,4]
push!(x, my_list)
end
```
vs
```
map(x-> x, [1,2,3,4])
```

[the remark this answers is not in the document]
Comment (charleskawczynski) [reply src/ClimaLSM.jl]: Yeah, IMO, the issue with `push!`ing is that it dynamically allocates on every `push!` call, whereas, IIUC, `map` will do a single allocation. If you know what the elements are:

```julia
my_list = Vector{Float64}(undef)
for i in (1,2,3,4)
    my_list[i] = ...
end
```
Then this also allocates once at `my_list = Vector{Float64}(undef)`.

[the remark this answers is not in the document]
Comment (kmdeck) [reply src/ClimaLSM.jl]: thanks!
